### ISSUE TITLE
tzwinlocal() does not work with non-ASCII characters in timezone name

### ISSUE THREAD

**Issue by pganssle** [dateutil/dateutil#210]
First reported as a comment on #197 by @labrys, this is actually unrelated to that issue, so it's getting its own issue:

> since updating to dateutil 2.5.0 in sickrage, windows users are getting this (possibly related) timezone error.  It seems to be a unicode related bug as its causing:
> ```python
>     sb_timezone = tz.tzwinlocal() if tz.tzwinlocal else tz.tzlocal()
>   File "C:\SickRage\SickRage\lib\dateutil\tz\win.py", line 239, in __init__
>     handle, "%s\%s" % (TZKEYNAME, self._stdname)) as tzkey:
> UnicodeEncodeError: 'ascii' codec can't encode character u'\xe4' in position 57: ordinal not in range(128)"
> ```
> The limited tests we've tried so far do not cause this error with `tz.tzlocal()`.  Changing timezones to e.g EST or to some other european timezones do not cause the error with `tz.tzwinlocal()`.
> This was confirmed using python 2.7.10 (portable) on Server 2012 R2 and Windows 7 ultimate.

I was able to confirm this issue myself. I tried a first pass mitigation in PR #208 (which fixed things on our end), but it turns out that this is a bug in Python 2.x where the `_winreg` module does not properly handle unicode.

Ideally this would be handled upstream, but we may be able to come up with a mitigation by retrieving the key another way. I'm not 100% sure, but given that there are apparently [canonical time zone keys](http://stackoverflow.com/a/32974430/467366) in English on Windows, maybe we can try and retrieve the time zone that way on Python 2.7.

Alternately, we can maybe use an approach similar to what [`winreg_unicode`](https://github.com/DanielStutzbach/winreg_unicode) is doing (I'd rather not add winreg_unicode as a dependency if it can be helped).

**Comment by labrys** [comment]
@pganssle thanks for creating the issue for me

**Comment by Prinz23** [comment]
Replying to: https://github.com/dateutil/dateutil/issues/434#issuecomment-320331315

You need the english name because the regkey has always a english name, the localized is for displaying only.

Currently `winreg` tires to open because it uses the localized `keydict["StandardName"]`:

`SOFTWARE\Microsoft\Windows NT\CurrentVersion\Time Zones\Mitteleuropäische Zeit`

but that regkey doesn't exists. The name is always english `keydict["TimeZoneKeyName"]` (Windows 7+): 
`SOFTWARE\Microsoft\Windows NT\CurrentVersion\Time Zones\W. Europe Standard Time`

so there is no unicode issue even, so there is no real `winreg` issue, since we never need to open a unicode regkey (because it doesn'r exists)

You can keep the local name in `self._std_abbr` and just change:

```python
tzkeyname = text_type('{kn}\\{sn}').format(kn=TZKEYNAME,
                                                          sn=self._std_abbr)
```
to:
```python
tzkeyname = text_type('{kn}\\{sn}').format(kn=TZKEYNAME,
                                                          sn=keydict["TimeZoneKeyName"])
```
For pre Windows 7 you need to find the english name just like in the linked tzlocal file by looping trough all `SOFTWARE\Microsoft\Windows NT\CurrentVersion\Time Zones\`  subkeys and compare the `Std` value to the `keydict["StandardName"]` value.

**Comment by dagwieers** [comment]
Is someone working on a fix ? We're three years into this now and I found out we're affected :-(
I guess trying to get this fixed for Python 2.7 has become impossible, yet affects us.
Working around it in dateutil (as described above) is the next best thing.

**Comment by pganssle** [comment]
@dagwieers As far as I know no one is working on it, feel free to pick up this issue.

**Comment by dagwieers** [comment]
@pganssle Is the [proposed solution from this thread](https://github.com/dateutil/dateutil/issues/210#issuecomment-320348647) as in #912 acceptable ? I genuinely wonder if we need to retain compatibility with pre-Windows 7.

**Comment by labrys** [comment]
@dagwieers in my opinion as long as there's a current Python version that supports the windows version it should still be supported.  For example after python 2.7 has reached end of life next year, XP support should be dropped.

**Comment by dagwieers** [comment]
@labrys Is there a link with information on what Windows versions are supported by each Python version ? I couldn't find anything using Google or in the Python documentation.

**Comment by labrys** [comment]
@dagwieers As of right now the only defining line for Windows version support is if its Windows XP and eaerlier.  The only current version of Python that still supports XP and earlier is Python 2.7.

The current list of Windows releases can be found at https://www.python.org/downloads/windows/
and releases that no longer support XP have a note in bold to that effect.

**Comment by dagwieers** [comment]
Well, it's one thing to know that (for technical reasons?) a specific Python version no longer works for a specific Windows version, but that doesn't really give us insight into what Windows versions are tested and supported.

Does the Windows XP statement suggest that Windows Vista and Windows Server 2003 are still tested and supported by Python 3?

**Comment by labrys** [comment]
Supported != Tested

I would say its a safe bet that Windows Vista and Server 2003 are not actively tested.

However, if a bug was reported on those platforms and someone provided a suitable fix I would expect it to be included until it is decided that the platform is no longer supported. 

While I don't expect the python team is actively developing for those systems, without a statement of exactly what is and is not supported, that link is the best we have for an official stance on version support.

**Comment by dagwieers** [comment]
> However, if a bug was reported on those platforms and someone provided a suitable fix I would expect it to be included until it is decided that the platform is no longer supported.

Right, I was hoping to find any formalized decision, rather than "as long as it works", or "nobody reported otherwise".

Especially since we end up with the lowest common denominator in the implementation when using available interfaces.
